### PR TITLE
matplotlib: declare missing runtime packages

### PR DESCRIPTION
**Pull request by joelmo**
###### Things done:
- [ x ] Tested using sandboxing (`nix-build --option build-use-chroot true` or [nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS)
- [ ] Built on platform(s): NixOS / OSX / Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ x ] Tested execution of all binary files (usually in `./result/bin/`)
- [ x ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).
###### More

Fixes issue #<insert id>

cc @<maintainer>

---

_Please note, that points are not mandatory, but rather desired._

### CODE REVIEW

**Comment by jagajaga** [comment]
What about Travis error?

**Comment by joelmo** [comment]
I will fix it soon

**Comment by joelmo** [comment]
Also tested with python3 now

**Comment by FRidh** [comment]
When did you notice these dependencies are missing? I haven't encountered any issue with missing dependencies but then again, matplotlib is big and I'm only using a part of it.

File: pkgs/development/python-modules/matplotlib/default.nix [hunk 1]
```diff
@@ -35,6 +35,11 @@ buildPythonPackage rec {
     ++ stdenv.lib.optional enableGtk2 pygtk
     ++ stdenv.lib.optionals enableGtk3 [ cairo pycairo gtk3 gobjectIntrospection pygobject3 ];
 
+  propagatedUserEnvPkgs =
+    [ pyparsing cycler dateutil
+    ] 
+    ++ (with stdenv.lib; optional ((hasAttr "isPy2" python) && python.isPy2) six);
```
Comment (FRidh) [inline pkgs/development/python-modules/matplotlib/default.nix]: please pass in `isPy3k` instead

**Comment by joelmo** [comment]
I use python 2. How do you have it installed? I have it installed with nix-env and then $PYTHONPATH points to that profile lib/python2..

File: pkgs/development/python-modules/matplotlib/default.nix (hunk 1 above)
Comment (joelmo) [inline pkgs/development/python-modules/matplotlib/default.nix]: Ok will change it

**Comment by FRidh** [comment]
That's why it does not work for you. Python libraries are not supposed to be installed with `nix-env`.
Instead, use `nix-shell` for your environments. Unfortunately this isn't explained in the manual yet but it should be hopefully soon: https://github.com/NixOS/nixpkgs/pull/13700

**Comment by FRidh** [comment]
Oh, and to get it working, try e.g.:
`nix-shell -p python27Packages.matplotlib python27Packages.ipython python27Packages.pytest`

**Comment by joelmo** [comment]
A shell like that pulls in build dependencies and I don't really need them. If I use a service for ipython notebook how do I use nix-shell to get the packages I want. I don't want them to the service.

**Comment by FRidh** [comment]
What type of packages do you want to add? Packages that are in Nixpkgs, or other ones, like those that you develop?

Is your intention to use an IPython Notebook server with a fixed set of packages? Or do you want to be able to quickly modify your package set and possibly use development/editable mode? 

If you want the latter (I use that a lot), then use `nix-shell`. You will have to write a Nix expression in that case.

If you want the former, then you might want to create a NixOS module. I have no experience with that, but I think @aragnon made one.

**Comment by joelmo** [comment]
Yes I want to make python packages from Nixpkgs available to use in a ipython notebook.

I want to be able to add/remove packages for this ipython environment without restarting the server or rebuilding nixos and rebooting. AFAIK there is no way other than using nix-env to solve this problem. 

And as I said before a nix-shell pulls in build dependencies, seems suboptimal?

**Comment by FRidh** [comment]
> Yes I want to make python packages from Nixpkgs available to use in a ipython notebook.

That's a very common use case. I'm using packages from Nixpkgs as well as some locally defined packages in combination with the notebook.

> I want to be able to add/remove packages for this ipython environment without restarting the server 

That is not going to be possible with Nix. With Nix you declare an environment along with a set of packages. If you work with Python you create an environment that has access to certain packages. This environment depends on those packages. If the set of packages changes. the environment changes and thus has to be rebuilt and the server restarted. What you describe can only be done in an impure way.

`nix-shell` pulls in what it needs to get you the environment you requested, just like any command that triggers `nix-build`. If you specify that you want an environment with a set of packages, then it only pulls in those packages and its run-time dependencies, but *not* its build-time dependencies. Unless of course the packages need to be build first, in which case you can garbage-collect the build-time dependencies after your packages have been build.

**Comment by joelmo** [comment]
> That is not going to be possible with Nix. With Nix you declare an environment along with a set of packages.

Now I am trying to understand why you think this isn't possible with Nix. A profile is an environment! So the interface between the server and environment have to be pure, why? This is a huge limitation if I have to rebuild and restart the server every time, what if I need to use data the process have created in memory, depending on what I do, saving that takes effort. Another concern, rebuilding services in nixos requires you to have root access.

If you count with propagatedBuildInputs, nix-shell pulls in build dependencies. I don't know if it should be counted.

I found how perl deals with runtime deps in nixpkgs, see  https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/perl-modules/generic/builder.sh.
This is like setting propagatedUserEnvPackages = propagatedBuildInputs, unless already defined. This should be done for python too.

**Comment by FRidh** [comment]
> This is a huge limitation if I have to rebuild and restart the server every time

You will have to rebuild your environment and restart your server when you change the packages it depends on.

> what if I need to use data the process have created in memory, depending on what I do, saving that takes effort. 

But why would you want to change the set of packages after it created data? You don't know on forehand which packages you need?

> Another concern, rebuilding services in nixos requires you to have root access.

Indeed. Therefore, why do you want to use this as a NixOS module, when you plan to change the environment constantly?

From what you told me it makes far more sense to write a `shell.nix` describing your environment, and run the notebook server from there.

**Comment by joelmo** [comment]
I think installing python packages using nix-env should be allowed. Not for use in development environments but more for cases where you use ipython and such software running as services. I don't have to restart the server when adding/removing packages. Privileged users that have access to that profile can manipulate it. Possibility to use NixOS service idiomatically (no rebuilds). A process started by a NixOS service can be different, like CPU quota and priority for processes can be set. And you have the possibility to setup the environment imperatively.

What do you think about setting propagatedUserEnvPackages = propagatedBuildInputs (unless propagatedUserEnvPackages already defined) for python modules? This is the way perl works in nix.

**Comment by domenkozar** [comment]
This is what Python also used to do until recent refactoring. 

We (the Python maintainers, mainly being me and @FRidh) decided not to maintain and support this feature, because there're lots of things to fix already and imperative package management is very limited.

Nevertheless, you should be possible to write a simple stdenv adapter to suit such needs. I'm +1 linking the github repository that supports this feature and willing to direct issues/questions there.

Note: you can still use `python.buildEnv` and using relative paths to $PREFIX should achieve the same effect as Nix profiles.

**Comment by joelmo** [comment]
How do I use `python.buildEnv`? I suppose the use is similar as `buildEnv`.

```
nix-repl> python.buildEnv { extraLibs = [ python2Packages.lxml ]; }
error: attempt to call something which is not a function but a set, at (string):1:1

nix-repl> python.buildEnv 
«derivation /nix/store/6awmwpdb5s36hpkb5783h4sc7hawnfyg-python-2.7.11-env.drv»
```

**Comment by domenkozar** [comment]
http://nixos.org/nixpkgs/manual/#ssec-python-build-env

**Pull request closed without merging**